contracts/adapters/GmpTransferAdapter.sol

```diff
@@ -0,0 +1,114 @@
+
+
+    event Withdrawal(address recipient, uint amount);
+    event FundsTransferred(address recipient, uint256 amount);
+
+    error SenderNotAdmin();
+    error InsufficientMsgValueAmount(uint256 amount);
+    error MsgValueLowerThanFee(uint256 amount);
+    error TokenWithdrawalFailed();
+    error InsufficientBalance();
+    error InvalidHandler(address handler);
+    error InvalidOriginAdapter(address adapter);
+    error FailedFundsTransfer();
+
+    error Invalid_MsgNotZero();
+    error FailedRefund();
+
+    modifier onlyAdmin() {
+        if (!hasRole(DEFAULT_ADMIN_ROLE, msg.sender)) revert SenderNotAdmin();
+        _;
+    }
+
+    /**
+        @notice This contract requires for transfer that the origin adapter address is the same across all networks.
+        Because of that it should be deployed using multichain deployer or create2.
+     */
+    constructor(IBridge bridge, address newGmpAddress, bytes32 resourceID) {
+        _bridge = bridge;
+        _gmpAddress = newGmpAddress;
+        _resourceID = resourceID;
+        _setupRole(DEFAULT_ADMIN_ROLE, msg.sender);
+    }
+
+    function deposit(uint8 destinationDomainID, address recipientAddress, address XERC20Address, uint256 tokenAmount) external payable {
+        address feeHandlerRouter = _bridge._feeHandler();
+        (uint256 fee, ) = IFeeHandler(feeHandlerRouter).calculateFee(
+            address(this),
+            _bridge._domainID(),
+            destinationDomainID,
+            _resourceID,
+            "", // depositData - not parsed
+            ""  // feeData - not parsed
+        );
+
+        if (msg.value < fee) {
+            revert InsufficientMsgValueAmount(msg.value);
+        // refund excess msg.value
+        } else if (msg.value > fee) {
+            (bool success, ) = msg.sender.call{value: msg.value - fee}("");
+            if (!success) revert FailedRefund();
+        }
+
+        bytes memory depositData = abi.encodePacked(
+            // uint256 maxFee
+            uint256(950000),
+            // uint16 len(executeFuncSignature)
+            uint16(4),
+            // bytes executeFuncSignature
+            IXERC20(address(0)).mint.selector,
+            // uint8 len(executeContractAddress)
+            uint8(20),
+            // bytes executeContractAddress
+            XERC20Address,
+            // uint8 len(executionDataDepositor)
+            uint8(20),
+            // bytes executionDataDepositor
+            address(this),
+            // bytes executionDataDepositor + executionData
+            prepareDepositData(recipientAddress, XERC20Address, tokenAmount)
+        );
+
+        IXERC20(XERC20Address).burn(msg.sender, tokenAmount);
+
+        _bridge.deposit{value: fee}(destinationDomainID, _resourceID, depositData, "");
+    }
+
+    function executeProposal(address gmpAdapter, address recipient, address XERC20Address, uint256 amount) external {
+        if (gmpAdapter != address(this)) revert InvalidOriginAdapter(gmpAdapter);
```

Comment:
Won't the origin adapter have a different address then the destination one?

Reply:
The plan is to deploy this with multichain deploy and have the same address across all supported networks

Reply:
Who will become the owner of the contract then?

Reply:
If I recall correctly there are 2 options, one is the address of the "deployer" of multichin deployment and for the other you can choose whatever address.

Reply:
It looks like we have Access control, but no admin methods. 
Might as well remove that.